src/app/CASESessionManager.cpp

```diff
@@ -43,19 +43,25 @@ CHIP_ERROR CASESessionManager::FindOrEstablishSession(PeerId peerId, Callback::C
 
         if (session == nullptr)
         {
-            onFailure->mCall(onFailure->mContext, peerId, CHIP_ERROR_NO_MEMORY);
-            return CHIP_ERROR_NO_MEMORY;
+            if (onFailure != nullptr)
+            {
+                onFailure->mCall(onFailure->mContext, peerId, CHIP_ERROR_NO_MEMORY);
+            }
+            return;
         }
     }
 
-    CHIP_ERROR err = session->Connect(onConnection, onFailure);
-    if (err != CHIP_NO_ERROR)
+    session->Connect(onConnection, onFailure);
+    if (!session->IsConnected() && !session->IsConnecting() && !session->IsResolvingAddress())
     {
-        // Release the peer rather than the pointer in case the failure handler has already released the session.
+        // This session is not making progress toward anything.  It will have
+        // notified the consumer about the failure already via the provided
+        // callbacks, if any.
+        //
+        // Release the peer rather than the pointer in case the failure handler
+        // has already released the session.
         ReleaseSession(peerId);
```

Comment:
Right now, the cleanup bits are .. very broken.  I don't think this PR makes them worse, though....  For example, bindings will detect that the pool filled up (where it can fill up) and then start evicting things....

I did consider interposing a thing here, which would be annoying and as you say not work if people directly hand callbacks to the OperationalDeviceProxy (which they are not supposed to do!) or just having an explicit backpointer from the OperationalDeviceProxy to the CASESessionManager.

Reply:
https://github.com/project-chip/connectedhomeip/issues/18609 tracks this.